hash/hash_comparator.go

```diff
@@ -1,55 +1,169 @@
 package hash
 
 import (
+	"bytes"
 	"context"
+	"crypto/aes"
+	"crypto/cipher"
 	"crypto/subtle"
 	"encoding/base64"
+	"encoding/hex"
 	"fmt"
 	"regexp"
 	"strings"
 
 	"github.com/pkg/errors"
 	"golang.org/x/crypto/argon2"
 	"golang.org/x/crypto/bcrypt"
+	"golang.org/x/crypto/sha3"
 
 	"github.com/ory/kratos/driver/config"
 )
 
-var ErrUnknownHashAlgorithm = errors.New("unknown hash algorithm")
+var ErrUnknownHashAlgorithm = fmt.Errorf("unknown hash algorithm")
+var ErrUnknownHashFormat = fmt.Errorf("unknown hash format")
+var ErrEmptyHashCompare = fmt.Errorf("empty hash provided")
+var ErrEmptyPasswordCompare = fmt.Errorf("empty password provided")
 
-func Compare(ctx context.Context, password []byte, hash []byte) error {
-	if IsBcryptHash(hash) {
-		return CompareBcrypt(ctx, password, hash)
-	} else if IsArgon2idHash(hash) {
-		return CompareArgon2id(ctx, password, hash)
-	} else {
+var hashSeparator = []byte("$")
+var bcryptLegacyPrefix = regexp.MustCompile(`^2[abzy]?$`)
+
+func Compare(ctx context.Context, cfg *config.Config, password []byte, hash []byte) error {
+	algorithm, realHash, err := ParsePasswordHash(hash)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	switch {
```

Comment:
Second guessing my previous thought we probably want to limit the amount of hashing methods to bcrypt family and argon2 for OS or highly confidential workloads. I think the current wrapping is fine and we probably want to have an onion model if we want to support more hash types for importing.

Reply:
I agree. We would probably need more flexible solution here so when dealing with very old systems that grew a number of hasher along the way it would not impact performance too much.